packages/flutter/lib/src/material/material_state.dart

```diff
@@ -407,6 +407,97 @@ abstract class MaterialStateOutlinedBorder extends OutlinedBorder implements Mat
   /// widget or theme.
   @override
   OutlinedBorder? resolve(Set<MaterialState> states);
+
+  /// Creates a [MaterialStateOutlinedBorder] from a
+  /// [MaterialPropertyResolver<OutlinedBorder?>] callback function.
+  ///
+  /// If used as a regular [OutlinedBorder], the border resolved in the default state
+  /// (the empty set of states) will be used.
+  ///
+  /// Usage:
+  ///
+  /// ```dart
+  /// ChipTheme(
+  ///   data: Theme.of(context).chipTheme.copyWith(
+  ///     shape: MaterialStateOutlinedBorder.resolveWith((Set<MaterialState> states) {
+  ///       if (states.contains(MaterialState.selected)) {
+  ///         return const StadiumBorder(
+  ///           side: BorderSide(color: Colors.transparent),
+  ///         );
+  ///       }
+  ///       return null;  // Defer to default value on the theme or widget.
+  ///     }),
+  ///   ),
+  ///   child: const Chip(
+  ///     label: Text('Transceiver'),
+  ///   ),
+  /// ),
+  /// ```
+  ///
+  /// Alternatively:
+  ///
+  /// ```dart
+  /// Chip(
+  ///   label: const Text('Transceiver'),
+  ///   shape: MaterialStateOutlinedBorder.resolveWith((Set<MaterialState> states) {
+  ///     if (states.contains(MaterialState.selected)) {
+  ///       return const StadiumBorder(
+  ///         side: BorderSide(color: Colors.transparent),
+  ///       );
+  ///     }
+  ///     return null;  // Defer to default value on the theme or widget.
+  ///   }),
+  /// ),
+  /// ```
+  static MaterialStateOutlinedBorder resolveWith(MaterialPropertyResolver<OutlinedBorder?> callback) =>
+      _MaterialStateOutlinedBorder(callback);
+}
+
+/// A [MaterialStateOutlinedBorder] created from a
+/// [MaterialPropertyResolver<OutlinedBorder>] callback alone.
+///
+/// If used as a regular [OutlinedBorder], the shape resolved in the default state will
+/// be used.
+///
+/// Used by [MaterialStateOutlinedBorder.resolveWith].
+class _MaterialStateOutlinedBorder extends MaterialStateOutlinedBorder {
+  const _MaterialStateOutlinedBorder(this._resolve);
+
+  final MaterialPropertyResolver<OutlinedBorder?> _resolve;
+
+  @override
+  OutlinedBorder? resolve(Set<MaterialState> states) {
+    return _resolve(states);
+  }
+
+  @override
+  OutlinedBorder copyWith({BorderSide? side}) {
```

Comment:
Why are all of the remaining methods unsupported? Unless there's some quirk that I'm missing this class should be similar to the other resolveWith classes in this file.

Reply:
Will those methods ever be used? If you use a `MaterialStateOutlinedBorder` in a widget that supports material states, the widget will always get the actual `OutlinedBorder` instance by calling the `resolve` method.

For example, the `Chip` widget:
https://github.com/flutter/flutter/blob/7c5481caa8480ad1993c2a3e944ec419acf00bb0/packages/flutter/lib/src/material/chip.dart#L981-L985

If the widget does not support material states, you probably should not be using a `MaterialStateOutlinedBorder`, because the behaviour won't be correct (that's why I added the `UnsupportedError`). Should you?

If we want to avoid throwing the error, one option is to use `resolve(<MaterialState>{})` in those methods. However, `resolve` can return `null`. What do I return if `null` is returned, I cannot get the default value from the theme there. E.g.:

```dart
@override
ShapeBorder scale(double t) {
  return resolve(<MaterialState>{})?.scale(t) ?? whatDoWeReturnHere;
}
```

Another option could be to make `MaterialStateOutlinedBorder` extend an actual implementation of an `OutlinedBorder`, like `StadiumBorder` so that those methods are already implemented (that's how `MaterialStateBorderSide` works because `BorderSide` is not abstract). However, the user may be using a different type of `OutlinedBorder`.

And one last option would be to accept a default value in the `MaterialStateOutlinedBorder` constructor. 

Any thoughts on how to proceed?

Reply:
It looks like we've already established a pattern for OutlinedBorder resolveWith methods in #91762, see https://api.flutter.dev/flutter/material/MaterialStateOutlineInputBorder/resolveWith.html

In this existing case the static resolve method was added to a concrete subclass.

Reply:
As you pointed out, I've used the `LinearBorder` concrete subclass, which removes the need to implement the abstract methods from `OutlinedBorder`.

I've also migrated the PR from `MaterialState` to `WidgetState`.